redhat-appstudio/service-provider-integration-operator
# feat: improved gitlab probing

Signed-off-by: Pavol Baran <pbaran@redhat.com>

### What does this PR do?
The GitLab Probe and its Examine method needs to know about every service provider configuration and based on that decide whether repoUrl is from GitLab.

In order minimalize the calls to k8s API and not complicate the Examine method, this listing is done prior to calling the Examine method which is done for every service provider.

### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

Solves https://issues.redhat.com/browse/SVPI-256

### How to test this PR?

1. Deploy SPI with operator image `quay.io/pbaran/spio:SVPI-256-probing`
2. Create user config secret such as `/samples/sp-secret.yaml` with a labes `spi.appstudio.redhat.com/service-provider-type: GitLab`  and `spi.appstudio.redhat.com/service-provider-host: some.gitlab.url.com`. The clientId and clientSecret should not matter. 
3. Create 3 bindings (with different names) such as `samples/binding-write-repo-gitlab.yaml` with different repo URLs: 
- https://some.gitlab.url.com/path/to/repo
- https://gitlab.com/path/to/repo
- https://gitlab.url.com/path/to/repo
4. The first two bindings should be linked to GitLab SpiAccessToken with their respective service provider URLs. The last should be linked to SpiAccessToken with HostCredentials service provider, as it does not match any known GitLab configuration.

5. IMPORTANT ADDITION. this PR now changes the shared code for service providers. Therefore you should also test creating GitHub and Quay bindings. You can use `samples/binding-write-repo.yaml` for GH and `samples/binding-quay-without-protocol.yaml` for Quay.

If there is a request to somehow limit this PR only for GitLab and introduce this change for every SP later, I am able to grant this.
